core/controllers/ingest.py

```diff
@@ -313,3 +316,17 @@ def save_submission_file(excel_file, submission_id):
     submission.submission_file = excel_file.stream.read()
     db.session.add(submission)
     db.session.commit()
+
+
+def remove_unreferenced_organisations():
+    """Removes organisations no longer referenced by a Programme.
+
+    Some organisation names have been wrong in previous ingest.
+    This will clean them up everytime a new ingest is run updating those names
+    """
+
+    organisations_without_child = Organisation.query.filter(~Organisation.programmes.any()).all()
```

Comment:
I think this can be optimised further. Using list comprehensions to unpack attributes of a list of results returned is potentially expensive as:

- The original query returns ALL the organisation objects/attributes, not just the attributes we need(id only)
- The list comp then does a lot more in local memory.

This can be improved by using the with_entities method. Also, I think we might need the scalars method to put them all in a list (rather than a list of tuples only containing id in each). Now you SHOULD get a lsit of just the id's, with all the expense done on DB server side

`organisation_ids_without_child = Organisation.query.filter(~Organisation.programmes.any()).with_entities(Organisation.id).scalar()`

(I haven't tested the above at all).

Reply:
I think I've found a better solution. .with_entities() seems to return the memory address, and scalar() only retrieves the first object. Instead we can just call all() at the end, and then if there's any there delete them (without needing to look up if a particular attribute is in a particular list.